src/backend/distributed/utils/maintenanced.c

```diff
@@ -257,6 +263,13 @@ CitusMaintenanceDaemonMain(Datum main_arg)
 		CHECK_FOR_INTERRUPTS();
 
 		/*
+		 * We clear the metadata cache before every iteration because otherwise
+		 * it might contain stale OIDs if deadlock detection runs during a DROP
+		 * EXTENSION followed by a CREATE EXTENSION.
+		 */
+		ClearMetadataCache();
```

Comment:
> The problem is that CitusHasBeenLoaded() may get the old pg_dist_partition OID from get_relname_relid after it has already been invalidated. 

With the locking I've prototyped? The `LockDatabaseObject()` does an `AcceptSharedInvalidationMessages()`, which'll trigger all pending invalidations. And as that's done before  `CitusHasBeenLoaded()` I can't see how this happens?  Or are you alleging some as of yet undiagnosed citus independent bug in cache invalidation in postgres?